### PR TITLE
Convert AbstractArrays with `strides` to NumPy arrays

### PR DESCRIPTION
**Pull request by mkitti**
`StridedArray`s can currently be converted to NumPy arrays with PyCall.jl. However, other types of `AbstractArray` also have a `stride` method. In particular, views such as `SubArray`, `ReshapedArray`, `ReinterpretArray`, and `PermutedDimsArray` implement `stride` but are not included in `StridedArray`.

This pull request:
1. Widens `PyCall.NpyArray` to accept any `AbstractArray`
2. If the array is immutable, embeds a reference to the parent array in the `PyObject`.
3. <strike>To support slicing, a new type `PySlice` is created.</strike>

### CODE REVIEW

**Comment by stevengj** [comment]
> Julia's PermutedDimsArray and StrideSubArray can transmitted to Python/Numpy without copying by taking advantage of numpy.transpose and Python's slice. The strategy is to create a view of the parent array and then use Numpy to perform similar view based operations on the array as was done in Julia

That shouldn't be necessary.  You should be able to create a NumPy array directly from any array with strided data, just by broadening the signature of [this method](https://github.com/JuliaPy/PyCall.jl/blob/318c23f3323d859b37ab4682339c03cb339e9135/src/numpy.jl#L175-L187).

**Comment by mkitti** [comment]
> > Julia's PermutedDimsArray and StrideSubArray can transmitted to Python/Numpy without copying by taking advantage of numpy.transpose and Python's slice. The strategy is to create a view of the parent array and then use Numpy to perform similar view based operations on the array as was done in Julia
> 
> That shouldn't be necessary. You should be able to create a NumPy array directly from any array with strided data, just by broadening the signature of [this method](https://github.com/JuliaPy/PyCall.jl/blob/318c23f3323d859b37ab4682339c03cb339e9135/src/numpy.jl#L175-L187).

Great. That will make things easier. I think we will need to modify some additional methods to dispatch to that method.

I'm not sure if I understand the dispatching with the code below in conversions.jl. It looks like we should be able to direct any `AbstractArray` with the method `stride` to the code you mention. Currently, such an array gets dispatched to:
* `PyObject(A::AbstractArray)`
* `array2py(A::AbstractArray)`
* `array2py(A::AbstractArray{<:Any, N}, dim::Integer, i::CartesianIndex{N})`

https://github.com/JuliaPy/PyCall.jl/blob/318c23f3323d859b37ab4682339c03cb339e9135/src/conversions.jl#L313-L333

Rather this needs to be directed to the `NpyArray` method you mentioned.

https://github.com/JuliaPy/PyCall.jl/blob/318c23f3323d859b37ab4682339c03cb339e9135/src/numpy.jl#L175-L195

Perhaps we should dispatch using a Holy trait:
* `PyObject(A::AbstractArray)`
* `py2array(A::AbstractArray, isstrided::Val{true})`
* `NpyArray(A::AbstractArray)`

Basically, this would mean that PyCall would try to convert any `AbstractArray` that has the method `strided` to a NumPy array rather than relying on `StridedArray` to dispatch.

Is changing `PyObject` dispatch in this way what you had in mind?

**Comment by stevengj** [comment]
I'm not sure we want to introduce a trait.   Just changing [this signature](https://github.com/JuliaPy/PyCall.jl/blob/318c23f3323d859b37ab4682339c03cb339e9135/src/numpy.jl#L189-L195) to `PyObject(a::AbstractArray{T}) where T<:PYARR_TYPES` (and similarly for `NpyArray `, which should probably be `npyarray`) should duck-type it — any type that doesn't have `strides` and a pointer conversion will fall back to `array2py`.

**Comment by mkitti** [comment]
I was trying to sort out the `Core.TypeofVararg` (see https://github.com/JuliaLang/julia/pull/38136 ) issue with the test on line 53. Turns out that the AOT CI script needing an overhaul now that some issues have been resolved. Previously it was always testing master.

File: src/conversions.jl [hunk 1]
```diff
@@ -751,6 +789,8 @@ function pysequence_query(o::PyObject)
         return typetuple(pytype_query(PyObject(ccall((@pysym :PySequence_GetItem), PyPtr, (PyPtr,Int), o,i-1)), PyAny) for i = 1:len)
     elseif pyisinstance(o, pyxrange[])
         return AbstractRange
+    elseif pyisinstance(o, pyslice[])
+        return PySlice
```
Comment (stevengj) [inline src/conversions.jl:793]: We don't default to creating no-copy wrappers of other Python sequences (e.g. lists, numpy arrays), no probably slices should be no different.

File: src/gc.jl [hunk 2]
```diff
@@ -29,7 +29,13 @@ const weakref_callback_meth = Ref{PyMethodDef}()
 function pyembed(po::PyObject, jo::Any)
     # If there's a need to support immutable embedding,
     # the API needs to be changed to return the pointer.
-    isimmutable(jo) && throw(ArgumentError("pyembed: immutable argument not allowed"))
+    if isimmutable(jo)
+        if hasmethod(parent, ( typeof(jo), ) )
```
Comment (stevengj) [inline src/gc.jl:33]: `applicable(parent, jo)`?

File: src/numpy.jl [hunk 3]
```diff
@@ -225,3 +225,23 @@ PyObject(a::Union{LinearAlgebra.Adjoint{<:Real},LinearAlgebra.Transpose}) =
...
 
 PyObject(a::LinearAlgebra.Adjoint) = PyObject(Matrix(a)) # non-real arrays require a copy
+
+# Alternative conversion of array views mapping parent in Julia to base in Numpy
+PyObjectWithParent(a::AbstractArray) = PyObject(a)
+
+function PyObjectWithParent(pda::PermutedDimsArray{T,N,perm}) where {T,N,perm}
+    parent = PyObjectWithParent(pda.parent)
+    # numpy.transpose is similar to PermutedDimsArray and creates a view of the data
+    pycall(parent.transpose, PyObject, perm .- 1)
```
Comment (stevengj) [inline src/numpy.jl:235]: What if `parent` is not a numpy array?  Why can't we create a numpy array directly from `pda` as discussed?

Comment (mkitti) [reply src/numpy.jl]: We can create a NumPy array directly from pda. I kept this vestigial code around for comparison. I have removed it now.

Here's an expanded version of what is currently in `runtests.jl`:

```julia
julia> A = Float64[1 2; 3 4]
2×2 Array{Float64,2}:
 1.0  2.0
 3.0  4.0

julia> B = PermutedDimsArray(A, (2,1) )
2×2 PermutedDimsArray(::Array{Float64,2}, (2, 1)) with eltype Float64:
 1.0  3.0
 2.0  4.0

julia> C = PyObject(B)
PyObject array([[1., 3.],
       [2., 4.]])

julia> D = PyArray(C)
2×2 PyArray{Float64,2}:
 1.0  3.0
 2.0  4.0

julia> D == B
true

julia> A[2,1] = 7
7

julia> A
2×2 Array{Float64,2}:
 1.0  2.0
 7.0  4.0

julia> B
2×2 PermutedDimsArray(::Array{Float64,2}, (2, 1)) with eltype Float64:
 1.0  7.0
 2.0  4.0

julia> C
PyObject array([[1., 7.],
       [2., 4.]])

julia> D
2×2 PyArray{Float64,2}:
 1.0  7.0
 2.0  4.0
```

File: src/gc.jl (hunk 2 above)
Comment (mkitti) [reply src/gc.jl]: Done in fa4ed28

File: src/conversions.jl (hunk 1 above)
Comment (mkitti) [reply src/conversions.jl]: I removed these two lines.

**Comment by mkitti** [comment]
@stevengj Let me know if you would like to split the `slice` and `PySlice` code into another PR. I could also rebase this if you would like.

**Comment by mkitti** [comment]
I removed the slice code from this pull request. If we're interested, I can submit that as a distinct pull request. The diff is now smaller.

File: src/conversions.jl (hunk 1 above)
Comment (mkitti) [reply src/conversions.jl]: I have now removed all the pyslice code from this pull request. It is not needed here.

**Comment by mkitti** [comment]
Hi @stevengj , would it be possible to merge this soon? I'm trying to decide whether I will be able to release a downstream package that uses this.

**Comment by PhilipVinc** [comment]
What’s the status on this? I would find it useful, so I would be willing to commit some time if necessary.

**Comment by mkitti** [comment]
Hi @stevengj , I was going through my old PRs, and I was wonder where we were on this pull request to expand no-copy conversion of Julia arrays to NumPy arrays? Is there anything you would like to change before merging?

**Comment by PhilipVinc** [comment]
@stevengj bump

**Comment by mkitti** [comment]
For those who need this sooner rather than later, here's a version I created for a [Discourse](https://discourse.julialang.org/t/converting-julia-arrays-views-to-numpy-arrays-via-pycall/61186) post:

```julia
julia> using PyCall

julia> A = rand(Int, 256);

julia> pytypeof( PyObject(reinterpret(UInt64, A)) )
PyObject <class 'list'>

julia> pytypeof( PyObject(@view(A[1:100])) )
PyObject <class 'list'>

julia> module PyCallHack
           import PyCall: NpyArray, PYARR_TYPES, @npyinitialize, npy_api, npy_type
           import PyCall: @pycheck, NPY_ARRAY_ALIGNED, NPY_ARRAY_WRITEABLE, pyembed
           import PyCall: PyObject, PyPtr
           const HACKED_ARRAYS = Union{SubArray{T}, Base.ReinterpretArray{T}, Base.ReshapedArray{T}, Base.PermutedDimsArray{T}} where T <: PYARR_TYPES
           function NpyArray(a::HACKED_ARRAYS{T}, revdims::Bool) where T <: PYARR_TYPES
               @npyinitialize
               size_a = revdims ? reverse(size(a)) : size(a)
               strides_a = revdims ? reverse(strides(a)) : strides(a)
               p = @pycheck ccall(npy_api[:PyArray_New], PyPtr,
                   (PyPtr,Cint,Ptr{Int},Cint, Ptr{Int},Ptr{T}, Cint,Cint,PyPtr),
                   npy_api[:PyArray_Type],
                   ndims(a), Int[size_a...], npy_type(T),
                   Int[strides_a...] * sizeof(eltype(a)), a, sizeof(eltype(a)),
                   NPY_ARRAY_ALIGNED | NPY_ARRAY_WRITEABLE,
                   C_NULL)
              return PyObject(p, a)
           end
           pyembed(po::PyObject, jo::HACKED_ARRAYS) = pyembed(po, jo.parent)
       end
Main.PyCallHack

julia> pytypeof( PyObject(reinterpret(UInt64, A)) )
PyObject <class 'numpy.ndarray'>

julia> pytypeof( PyObject(@view(A[1:100])) )
PyObject <class 'numpy.ndarray'>
```

**Comment by PhilipVinc** [comment]
@stevengj bump

**Comment by mkitti** [comment]
I've initiated the registration of https://github.com/mkitti/NumPyArrays.jl which brings most of the functionality of this PR to a new type, `NumPyArray`. `NumPyArray` is just a wrapper of `PyArray` but has a constructor for an `AbstractArray`.

```julia
julia> using PyCall, NumPyArrays

julia> rA = reinterpret(Int8, rand(UInt8, 4,4))
4×4 reinterpret(Int8, ::Matrix{UInt8}):
 115   -4  -20   -32
 -43   13  -62   -36
 -84  -37   91  -109
 114  -40    9    73

julia> pytypeof(PyObject(rA))
PyObject <class 'list'>

julia> nprA = NumPyArray(rA)
4×4 NumPyArray{Int8, 2}:
 115   -4  -20   -32
 -43   13  -62   -36
 -84  -37   91  -109
 114  -40    9    73

julia> pytypeof(nprA)
PyObject <class 'numpy.ndarray'>

julia> sum(nprA)
-52

julia> pointer(rA) == pointer(nprA)
true

julia> np = pyimport("numpy")
PyObject <module 'numpy' from 'C:\\Users\\kittisopikulm\\.julia\\conda\\3\\lib\\site-packages\\numpy\\__init__.py'>

julia> np.sum(nprA)
-52

julia> PyObject(rA).dtype
ERROR: KeyError: key :dtype not found
Stacktrace:
 [1] __getproperty(o::PyObject, s::Symbol)
   @ PyCall ~\.julia\packages\PyCall\BD546\src\PyCall.jl:307
 [2] getproperty(o::PyObject, s::Symbol)
   @ PyCall ~\.julia\packages\PyCall\BD546\src\PyCall.jl:312
 [3] top-level scope
   @ REPL[33]:1

julia> PyObject(nprA).dtype
PyObject dtype('int8')

julia> py"""
       def add_one(A):
           return A + 1

       """

julia> py_add_one = py"add_one"
PyObject <function add_one at 0x00000000014BFF70>

julia> py_add_one(nprA)
4×4 Matrix{Int8}:
 116   -3  -19   -31
 -42   14  -61   -35
 -83  -36   92  -108
 115  -39   10    74

julia> py_add_one(rA)
ERROR: PyError ...

julia> py_add_one(PyObject(rA))
ERROR: PyError ...
```

**Comment by mkitti** [comment]
I intend on closing this pull request on January 11th.

**Comment by mkitti** [comment]
Closing this. See https://github.com/mkitti/NumPyArrays.jl if interested.
